src/ScheduleFunctions.cpp

```diff
@@ -874,6 +994,642 @@ class InjectRealization : public IRMutator2 {
+                             for_loop->min,
+                             for_loop->extent,
+                             for_loop->for_type,
+                             for_loop->device_api,
+                             body);
+        }
+    }
+};
+
+// Inject 'injected' into 'root' at 'level'.
+Stmt inject_stmt(Stmt root, Stmt injected, const LoopLevel &level) {
+    if (!root.defined()) {
+        return injected;
+    }
+    if (!injected.defined()) {
+        return root;
+    }
+    if (level.is_inline() || level.is_root()) {
+        return Block::make(root, injected);
+    }
+    InjectStmt injector(injected, level);
+    root = injector.mutate(root);
+    internal_assert(injector.found_level);
+    return root;
+}
+
+// Collect all let stmts that define the loop min, max, and extent.
+class CollectBounds : public IRVisitor {
+public:
+    map<string, Expr> bounds;
+
+private:
+    using IRVisitor::visit;
+
+    void visit(const LetStmt *op) {
+        if (ends_with(op->name, ".loop_min") ||
+            ends_with(op->name, ".loop_max") ||
+            ends_with(op->name, ".loop_extent")) {
+            bounds.emplace(op->name, Variable::make(Int(32), op->name));
+        }
+        IRVisitor::visit(op);
+    }
+};
+
+class SubstituteFusedBounds : public IRMutator {
+public:
+    const map<string, Expr> &replacements;
+    SubstituteFusedBounds(const map<string, Expr> &r) : replacements(r) {}
+
+private:
+    using IRMutator::visit;
+
+    void visit(const For *op) {
+        const Variable *min_var = op->min.as<Variable>();
+        const Variable *extent_var = op->extent.as<Variable>();
+        if (min_var && extent_var) {
+            Expr min_val, extent_val;
+            {
+                const auto &it = replacements.find(min_var->name);
+                if (it != replacements.end()) {
+                    min_val = it->second;
+                }
+            }
+            {
+                const auto &it = replacements.find(extent_var->name);
+                if (it != replacements.end()) {
+                    extent_val = it->second;
+                }
+            }
+
+            if (!min_val.defined()|| !extent_val.defined()) {
+                IRMutator::visit(op);
+                return;
+            }
+
+            Stmt body = mutate(op->body);
+
+            size_t last_dot = op->name.rfind('.');
+            internal_assert(last_dot != string::npos);
+            string new_var = op->name.substr(0, last_dot) + ".fused." + op->name.substr(last_dot + 1);
+
+            // If this is the child fused loop, might as well clear the for-loop
+            // scheduling flag (parallel, vectorize, or unrolled) since its
+            // loop extent is one anyway.
+            ForType for_type = op->for_type;
+            DeviceAPI device_api = op->device_api;
+            if (is_one(extent_val)) {
+                for_type = ForType::Serial;
+                device_api = DeviceAPI::None;
```

Comment:
Not at all obvious that this is safe. A GPUBlocks loop of extent one is different to a Hexagon loop of extent one is different to a serial loop of extent 1. I assume it's safe because the loops have already been checked for compatibility?